Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,593 @@
+of :term:`generators <generator iterator>` and
+:term:`generator functions <generator>`.
+Recall, a generator function is a function that :keyword:`yield`\s, like this
+one::
+
+   def get_random_number():
+       # This would be a bad random number generator!
+       print("Hi")
+       yield 1
+       print("Hello")
+       yield 7
+       print("Howdy")
+       yield 4
+       ...
+
+Similar to a coroutine function, calling a generator function does not run it.
+Instead, it creates a generator object::
+
+   >>> get_random_number()
+   <generator object get_random_number at 0x1048671c0>
+
+You can proceed to the next ``yield`` of a generator by using the
+built-in function :func:`next`.
+In other words, the generator runs, then pauses.
+For example::
+
+   >>> generator = get_random_number()
+   >>> next(generator)
+   Hi
+   1
+   >>> next(generator)
+   Hello
+   7
+
+=====
+Tasks
+=====
+
+Roughly speaking, :ref:`tasks <asyncio-task-obj>` are coroutines (not coroutine
+functions) tied to an event loop.
+A task also maintains a list of callback functions whose importance will become
+clear in a moment when we discuss :keyword:`await`.
+The recommended way to create tasks is via :func:`asyncio.create_task`.
+
+Creating a task automatically schedules it for execution (by adding a
+callback to run it in the event loop's to-do list, that is, collection of jobs).
+
+Since there's only one event loop (in each thread), :mod:`!asyncio` takes care of
+associating the task with the event loop for you. As such, there's no need
+to specify the event loop.
+
+::
+
+   coroutine = loudmouth_penguin(magic_number=5)
+   # This creates a Task object and schedules its execution via the event loop.
+   task = asyncio.create_task(coroutine)
+
+Earlier, we manually created the event loop and set it to run forever.
+In practice, it's recommended to use (and common to see) :func:`asyncio.run`,
+which takes care of managing the event loop and ensuring the provided
+coroutine finishes before advancing.
+For example, many async programs follow this setup::
+
+   import asyncio
+
+   async def main():
+       # Perform all sorts of wacky, wild asynchronous things...
+       ...
+
+   if __name__ == "__main__":
+       asyncio.run(main())
+       # The program will not reach the following print statement until the
+       # coroutine main() finishes.
+       print("coroutine main() is done!")
+
+It's important to be aware that the task itself is not added to the event loop,
+only a callback to the task is.
+This matters if the task object you created is garbage collected before it's
+called by the event loop.
+For example, consider this program::
+
+   async def hello():
+       print("hello!")
+
+   async def main():
+       hello_task = asyncio.create_task(hello())
+       return
+
+   asyncio.run(main())
+
+Because the coroutine ``main()`` exits before awaiting the task and no other
+references to the task are made, the task object ``hello_task`` *might* be
```

Comment:
This may not be the best example, as the application is expected to close at the end of main(). So the task gets cancelled anyway. i.e. Even if I keep a reference, I can get the task to not complete:
```
>>> async def hello():
...  await asyncio.sleep(1)
...  print("BAR")
... 
>>> asyncio.run(main())
>>> async def main():
...  global t
...  t = asyncio.create_task(hello())
... 
>>> asyncio.run(main())
>>> t
<Task cancelled name='Task-10' coro=<hello() done, defined at <stdin>:1>>
```

Reply:
> The risk that the current example is meant to emphasize is the problem with losing references (before `await`ing).

My point is that it doesn't show that, because main() exits regardless and the other task is not expected to continue running. As I showed above, it doesn't run past the initial iteration if we _do_ hold references to it, nevermind if we don't.

The second example has a sleep so that main() outlives the task. So, in this case the lack of a reference could cause an issue. With a reference, the task _is_ meant to run until completion.

Reply:
Yeah, that looks better to me.